src/Input/Tokenizer.php

```diff
@@ -0,0 +1,464 @@
+ * converts them into pre-defined tokens.
+ *
+ * @author  Shlomo Hassid <shlomohassid@gmail.com>
+ * @license MIT
+ *
+ * @link    https://github.com/adhocore/cli
+ */
+class Tokenizer implements Iterator
+{
+    
+    /** @var Token[]  */
+    private array $tokens = [];
+
+    private int $index = 0;
+
+    /**
+     * @param array $args
+     */
+    public function __construct(array $args) 
+    {
+        // Flags:
+        $variadic = false;
+        $literal  = false;
+
+        // Process args:
+        foreach ($args as $arg) {
+
+            // Tokenize this arg:
+            $tokens = [];
+            if (
+                // Not a literal token:
+                !$literal || 
+                // Or a literal token with variadic close:
+                // Which is a special case:
+                ($variadic && ($arg[0] ?? '') === Token::TOKEN_VARIADIC_C)
+            ) {
+                
+                $tokens = $this->tokenize($arg);
+                $literal = false;
+
+            } else {
+                // Literal token treat all as constant:
+                $tokens[] = new Token(Token::TYPE_CONSTANT, $arg);
+            }
+
+            // Process detected token/s:
+            foreach ($tokens as $token) {
+            
+                switch ($token->type()) {
+            
+                    case Token::TYPE_VARIADIC:
+                        $variadic = $token->isVariadic("open");
+                        if ($variadic) {
+                            $this->tokens[] = $token;
+                        }
+                        break;
+
+                    case Token::TYPE_LITERAL:
+                        // Add all remaining tokens as nested of a literal token:
+                        $literal = true;
+                        break;
+
+                    case Token::TYPE_SHORT:
+                    case Token::TYPE_LONG:
+                    case Token::TYPE_CONSTANT:
+                        if ($variadic) {
+                            $this->tokens[count($this->tokens) - 1]->addNested($token);
+                        } else {
+                            $this->tokens[] = $token;
+                        }
+                        break;
+                }
+            }
+        }
+    }
+
+    /**
+     * Get the detected tokens.
+     * 
+     * @return Token[]
+     */
+    public function getTokens(): array
+    {
+        return $this->tokens;
+    }
+
+    /**
+     * Detect constants: strings, numbers, negative numbers.
+     * e.g. string, 123, -123, 1.23, -1.23
+     * 
+     * @param string $arg
+     * 
+     * @return bool
+     */
+    private function isConstant(string $arg): bool
+    {   
+        // Early return for non-option args its a constant:
+        if (!$this->isOption($arg)) { 
+            return true;
+        }
+        // If its a single hyphen, maybe its a negative number:
+        if (
+            ($arg[0] ?? '') === '-'
```

Comment:
btw we cant assume `-N` where N is number, to be a number it can be a option as well

eg `curl` uses `-1` `-2` `-3` as option for ssl v1 to v3

Reply:
This is pretty rare and usually confusing, I can think of a way to support numerical names for options by checking first if the current identified number is defined as an option. 
That said, it still introduces some extra complexity and weirdness : -10 -12 ...